File: mailblastr-python/mailblastr/segments.py
"""Segments resource — DOMAIN-FIRST: segments belong to a sending domain.
``domain`` is REQUIRED on create and list; names are unique within a domain
(reusable across domains) and every domain carries an auto-created "General"
segment."""

from typing import Any, Dict, TypedDict

from . import http_client
from .http_client import build_query, paginate, path_escape as _e


class CreateParams(TypedDict, total=False):
    domain: str  # REQUIRED: the sending domain this segment belongs to
    name: str  # required
    # {status?, email_contains?, property_filters?, engagement?} — each key is
    # also accepted at the top level of the body, where `filter` wins.
    filter: Dict[str, Any]


class UpdateParams(TypedDict, total=False):
    # A segment's domain cannot change: sending `domain` (or `audience_id`) at
    # all is a 422 — create a new segment on the other domain instead.
    name: str
    filter: Dict[str, Any]


class ListParams(TypedDict, total=False):
    domain: str  # REQUIRED: the sending domain whose segments to list
    limit: int
    after: str
    before: str


class Segments:
    """``mailblastr.Segments`` — the /segments endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams
    ListParams = ListParams

    @classmethod
    def create(cls, params):
        """Create a segment on a sending domain (``domain`` is required).
        POST /segments"""
        return http_client.request("POST", "/segments", params)

    @classmethod
    def get(cls, segment_id):
        """Retrieve a segment. GET /segments/:id"""
        return http_client.request("GET", f"/segments/{_e(segment_id)}")

    @classmethod
    def list(cls, params):
        """List a domain's segments (``domain`` is required; includes the
        auto-created "General" segment). GET /segments?domain=..."""
        qs = build_query(
            {
                "domain": params["domain"],
                "limit": params.get("limit"),
                "after": params.get("after"),
                "before": params.get("before"),
            }
        )
        return http_client.request("GET", f"/segments{qs}")

    @classmethod
    def contacts(cls, segment_id, params=None):
        """Preview the contacts a segment currently resolves to (filter matches
        plus explicit memberships). Items are a reduced contact shape with no
        ``properties``. Called with no pagination params every contact is
        returned. GET /segments/:id/contacts"""
        return http_client.request(
            "GET", f"/segments/{_e(segment_id)}/contacts{paginate(params)}"
        )

    @classmethod
    def update(cls, segment_id, params):
        """Update a segment. PATCH /segments/:id"""
        return http_client.request("PATCH", f"/segments/{_e(segment_id)}", params)

    @classmethod
    def remove(cls, segment_id):
        """Delete a segment. DELETE /segments/:id"""
        return http_client.request("DELETE", f"/segments/{_e(segment_id)}")

File: mailblastr-java/src/main/java/com/mailblastr/resources/Segments.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;
import com.mailblastr.http.Query;
import com.mailblastr.requests.CreateSegmentRequest;
import com.mailblastr.requests.UpdateSegmentRequest;

/**
 * Segments (DOMAIN-FIRST) — {@code domain} is REQUIRED on create and list.
 * Segment names are unique within a domain; every domain carries an
 * auto-created "General" segment.
 */
public final class Segments extends Resource {
    public Segments(ApiClient api) { super(api); }

    /** {@code POST /segments} — {@code domain} is required on the request. */
    public MailblastrResponse create(CreateSegmentRequest request) {
        return api.request("POST", "/segments", request);
    }

    /** {@code GET /segments/:id} */
    public MailblastrResponse get(String id) {
        return api.request("GET", "/segments/" + enc(id));
    }

    /** List a domain's segments. {@code GET /segments?domain=} */
    public MailblastrResponse list(String domain) { return list(domain, null); }

    public MailblastrResponse list(String domain, ListParams params) {
        Query q = new Query().add("domain", domain);
        if (params != null) params.applyTo(q);
        return api.request("GET", "/segments" + q);
    }

    /** Preview the contacts a segment currently resolves to. {@code GET /segments/:id/contacts} */
    public MailblastrResponse contacts(String id) {
        return api.request("GET", "/segments/" + enc(id) + "/contacts");
    }

    /** {@code PATCH /segments/:id} */
    public MailblastrResponse update(String id, UpdateSegmentRequest request) {
        return api.request("PATCH", "/segments/" + enc(id), request);
    }

    /** {@code DELETE /segments/:id} */
    public MailblastrResponse remove(String id) {
        return api.request("DELETE", "/segments/" + enc(id));
    }
}
